\documentclass[11pt,a4paper]{article}
\usepackage{Sweave}
\usepackage{listings}
\usepackage[hidelinks]{hyperref}
\lstset{
   breaklines=true,
   basicstyle=\ttfamily}

\begin{document}
\SweaveOpts{concordance=TRUE}
\title{Computational Diagnostics:  Assignment 1}
\author{Mr Michael J. Jones \\
  \em MSc Bioinformatics and \\
  \em Computational Genomics}
\date{\today}
\maketitle

\tableofcontents

<<echo=FALSE>>=
options(width=60)
@

\section{Needleman and Wunsch Alignment}
The Needleman and Wunsch sequence alignment algorithm was implemented as an R package and tested thoroughly by use of the \verb@testthat@ unit testing framework and uploaded to \href{https://github.com}{github.com}.\footnote{The source-code can be freely downloaded at \url{https://github.com/hiraethus/Needleman-Wunsch}}

\subsection{Installation}
As the package uses Hadley Wickham's \verb@devtools@, the package can be easily installed. Note that this method will only work with distributions of R using Linux and Mac OS X.

<<>>=
#  only install hiraethus.needleman.wunsch if not already installed
if (!("hiraethus.needleman.wunsch" %in% installed.packages())) {
  devtools::install_github("hiraethus/Needleman-Wunsch")
}
# now load the library
library("hiraethus.needleman.wunsch")
@

\subsection{Demonstrating Needleman-Wunsch implementation}
The signature for the needleman-wunsch function is \begin{verbatim}function (seq1, seq2, gap = 0, match = 1, mismatch = 0).\end{verbatim}

We will compare the result of doing a sequence alignment on the two example sequences
\begin{itemize}
  \item $sequence1 \leftarrow ``\Sexpr{sequence1 <- "ACCCGGTCGTCAATTA"}$''
  \item $sequence2 \leftarrow ``\Sexpr{sequence2 <- "ACCACCGGTTGGTCCAATAA"}$''
\end{itemize}

using different parameters for the scoring 
function. Firstly, let us set the parameters as $match \leftarrow \Sexpr{match <- 1}$, $mismatch \leftarrow \Sexpr{mismatch <- -1}$ and $gap \leftarrow \Sexpr{gap <- -1}$.

<<>>=
first.alignment <- needle(sequence1, sequence2, gap, match, mismatch)
print(first.alignment)
@

The alignment yielded a near perfect alignment between the two sequences. Only one misalignment occurred.

If we change the parameters, say by making the mismatch penalty zero, we can still see that in this case, the alignments are the same but the maxscore is higher. This time, the parameters are $match \leftarrow \Sexpr{match <- 1}$, $mismatch \leftarrow \Sexpr{mismatch <- 0}$ and $gap \leftarrow \Sexpr{gap <- -1}$.

<<>>=
second.alignment <- needle(sequence1, sequence2, gap, match, mismatch)
print(second.alignment)
@

Again, we observe exactly the same alignment however, we get a different maxscore.

The traceback and score matrices are stored in the result object calculated by the \verb@needle@ function.

<<>>=
first.alignment.traceback <- first.alignment$traceback.matrix
second.alignment.traceback <- second.alignment$traceback.matrix
traceback.mismatches <- length(which(
  first.alignment.traceback != second.alignment.traceback))
print(traceback.mismatches)
@

Even though the two alignments returned were equal, the actual traceback matrices used to calculate the sequence alignments 
yield a number of different results as well. We can see from the above code that $\Sexpr{traceback.mismatches}$ mismatches 
are found between the two traceback matrices.

The score matrices used to calculate the $maxscore$ in our sequence alignment can be compared in the same way.
<<>>=
first.alignment.score <- first.alignment$score.matrix
second.alignment.score <- second.alignment$score.matrix
score.mismatches <- length(which(
  first.alignment.score != second.alignment.score))
print(score.mismatches)
@

In the case of the score matrices, we see that $\Sexpr{score.mismatches}$ mismatches were found.
<<echo=FALSE>>=
# clear env
rm(list=ls())
@

\section{Differential Gene Expression Analysis}

\subsection{Running the script}
% make sure computational-diagnostics-report-1 public before submit
The script for performing this element of the assignment can be downloaded from the url 
\begin{verbatim}
https://raw.githubusercontent.com/hiraethus/computational-diagnostics-report-1/
  master/q2_and_q3_expression_analysis.R
\end{verbatim}
and is also included in the appendix of this document.

<<echo=FALSE, print=FALSE, results=hide>>=
source("q2_and_q3_expression_analysis.R")
@
\subsection{Results}
\subsubsection*{How many probesets do you find differentially expressed using a false discovery rate (FDR) of 0.05?}
\Sexpr{num.diff.exp.probesets} are differentially expressed given an FDR of 0.05.

\subsubsection*{How many unique entrez gene identifiers are in the dataset and how many unique entrez identifiers are differentially expressed?}
\Sexpr{count.total.gene.ids} unique entrez gene identifiers are in the dataset and \Sexpr{count.diff.expressed.gene.ids} 
unique entrez identifiers are differentially expressed (given an FDR of 0.05). For details of how these were created, 
please see the script.

\subsubsection*{Show a table of the top 15 differentially expressed probesets along with their gene symbol, log fold change, average expression, t-score, p-value, adjusted p-value, and B statistic}

(For details of how this table was created, please see the script.)

<<top15,results=tex>>=
library(xtable)
xtab <- xtable(top.15.results, caption = "Top 15 results", label="tab:top15")
print(xtab)
@

% TODO: APPENDIX- Add all code at the end...will be horrible.
\end{document}
